### DISCUSSION TITLE
How do I build non-Nest.js internal packages and still get instant type-checking in VSCode?

### DISCUSSION THREAD

**Discussion by jtlapp** [vercel/turborepo#3640] [Help] [answered]
How do I transpile non-Next.js internal packages and still get instant type-checking?

The Next.js builder includes transpiling TypeScript, but I have a vanilla node app that also needs to include packages. If I set those packages up configured for Next.js, I get this error:

```
SyntaxError: Unexpected token 'export'
```

The [docs](https://turbo.build/repo/docs/handbook/sharing-code/internal-packages) suggest that the solution to this problem is to manually transpile the code. The [the kitchen sink example](https://github.com/vercel/turbo/blob/main/examples/kitchen-sink) solves the problem by [building a package](https://github.com/vercel/turbo/blob/main/examples/kitchen-sink/packages/logger/package.json) locally within the package directory.

However, using this solution, I have to build the package before any other package recognizes it or any change to it. This is a pain in the tush, an issue I'm not having with Next.js-only packages. When I make a change to a Next.js-only package, VSCode immediately recognizes those changes in all packages that import it, allowing for easy refactoring and use of new functions.

I attempted to use `tsup` to build my node app, hoping it would propagate transpilation through to dependent internal packages, but it did not. I used this command:

```
turbo build --filter installer
```

With this `apps/installer/package.json`:

```json
{
  "name": "installer",
  "version": "0.0.0",
  "private": true,
  "scripts": {
    "build": "tsup",
    "clean": "rm -rf .turbo && rm -rf node_modules",
    "dev": "tsup --watch",
    "lint": "TIMING=1 eslint \"src/**/*.ts*\"",
    "test": "jest --passWithNoTests"
  },
  "tsup": {
    "entry": [
      "src/main.ts"
    ],
    "splitting": false,
    "sourcemap": true,
    "clean": true,
    "external": [
      "pg-native"
    ]
  },
  "dependencies": {
    "@fieldzoo/config": "*",
    "@fieldzoo/database": "*",
    "@fieldzoo/utilities": "*"
  },
  "devDependencies": {
    "eslint-config-custom": "*",
    "jest-presets": "*",
    "tsconfig": "*"
  }
}
```

But I still get the "Unexpected token 'export'" error when I attempt to run the `installer` program.

How do I build non-Nest.js internal packages with transpilation and still get instant type-checking in VSCode?

**Comment by jtlapp** [comment, marked as answer]
I solved the problem, but I'm now wondering why the example repos aren't doing this. Just replace this:

`"types": "dist/index.d.ts"`

with this:

`"types": "src/index.ts"`

Here is the `package.json` for my non-Next.js app:

```json
{
  "name": "installer",
  "version": "0.0.0",
  "private": true,
  "scripts": {
    "build": "tsc",
    "clean": "rm -rf .turbo && rm -rf node_modules && rm -rf dist",
    "dev": "tsc -w",
    "lint": "TIMING=1 eslint \"src/**/*.ts*\"",
    "test": "jest --passWithNoTests"
  },
  "dependencies": {
    "@fieldzoo/config": "*",
    "@fieldzoo/database": "*",
    "@fieldzoo/utilities": "*"
  },
  "devDependencies": {
    "eslint-config-custom": "*",
    "jest-presets": "*",
    "tsconfig": "*"
  }
}
```

Here is the app's `tsconfig.json`:

```json
{
  "extends": "tsconfig/base.json",
  "compilerOptions": {
    "lib": ["ES2019"],
    "module": "commonjs",
    "outDir": "./dist",
    "rootDir": "./src"
  },
  "include": ["*.d.ts", "**/*.ts", "**/*.tsx"],
  "exclude": ["dist", "build", "node_modules"]
}
```

Here is an example `package.json` for an internal library:

```json
{
  "name": "@fieldzoo/validators",
  "version": "0.0.0",
  "private": true,
  "main": "dist/index.js",
  "source": "src/index.ts",
  "types": "src/index.ts",
  "files": [
    "dist/**"
  ],
  "scripts": {
    "build": "tsc",
    "clean": "rm -rf .turbo && rm -rf node_modules && rm -rf dist",
    "dev": "tsc -w",
    "lint": "TIMING=1 eslint \"src/**/*.ts*\"",
    "test": "jest"
  },
  "jest": {
    "preset": "jest-presets/node"
  },
  "devDependencies": {
    "eslint-config-custom": "*",
    "jest-presets": "*",
    "tsconfig": "*"
  }
}
```

And here is the library's `tsconfig.json`:

```json
{
  "extends": "tsconfig/lib.json",
  "compilerOptions": {
    "lib": ["ES2019"],
    "module": "commonjs",
    "outDir": "./dist",
    "rootDir": "./src"
  },
  "include": ["*.d.ts", "**/*.ts", "**/*.tsx"],
  "exclude": ["dist", "build", "node_modules"]
}
```

**Reply by jtlapp** [reply]
Will this solution be a problem in published packages?

**Reply by anthonyshew** [reply]
It has the potential to be problematic in a published package, yes.

If you're using `pnpm`, try out [the `publishConfig` property](https://pnpm.io/package_json#publishconfig) to override some of these values before your package is packed.

For other package managers, you'd have to configure your package for publishing and then wire it into your monorepo as if it were published.
